### ISSUE TITLE
cmd/govim: support temporary disabling of plugin/gopls

### ISSUE THREAD

**Issue by myitcv** [myitcv/govim#77]
Courtesy of @mvdan 

> Will there be a mode to "turn off", or otherwise pause/kill the lsp server? Use case; sometimes I benchmark stuff, so I don't want a random cpu spike to tamper with the results. And if the lsp periodically does work, or happens to parse or typecheck something, that could well happen

**Comment by mvdan** [comment]
It would be ideal if the answer was some sort of "pause all activity" instead of "kill the server and start it again later". The latter option would probably mean we'd lose much of the useful in-memory cache.

**Comment by myitcv** [comment]
cc @stamblerre @ianthehat for any thoughts

**Comment by ianthehat** [comment]
At the moment the gopls goes idle when you stop sending it changes. The time it takes to go idle however is very unpredictable, and when we start watching for file differences any change to a file it is watching will trigger it again.
The trouble is, if we stop replying to the editor eventually the rpc protocol backs up, so we would have to start returning errors instead, and then the editor might decide we are broken and restart us, which would be worse.
I cant think of anything we can reasonably do on this front, the solution has to be at the editor level not the plugin level.

**Comment by cespare** [comment]
After using govim for a while, I've run into a few rather unpleasant gopls bugs :( It seems like in general there are several cases where gopls can become confused about the state of the files and the only solution is to restart it. So I'll look into supporting this inside govim directly; probably pretty soon depending on how annoyed I get. (In addition to making govim handle gopls being killed and restarting it gracefully, I'll probably add a govim command to kill gopls.)

**Comment by myitcv** [comment]
> After using govim for a while, I've run into a few rather unpleasant gopls bugs

We should absolutely make sure we have bugs open in `gopls` for this, with tracking bugs here in `govim`. Are you comfortable they are covered elsewhere? If not, happy to ensure we have the right bugs raised.

> In addition to making govim handle gopls being killed and restarting it gracefully, I'll probably add a govim command to kill gopls.

Sounds good.

A command to shutdown/startup/restart `gopls` should be a relatively straightforward first step.

Having `govim` handle `gopls` dying is slightly more tricky because of the race with events arriving from Vim. Just as a thought, might be worth doing this as a second phase of implementation?

**Comment by cespare** [comment]
> We should absolutely make sure we have bugs open in `gopls` for this, with tracking bugs here in `govim`. Are you comfortable they are covered elsewhere?

One bug is #461. For the others, I have no repro steps yet. I've seen things similar to https://github.com/golang/go/issues/33608 and various other open and closed gopls bugs. One trigger was switching git branches, but I don't have a repeatable repro case.

> Having `govim` handle `gopls` dying is slightly more tricky because of the race with events arriving from Vim. Just as a thought, might be worth doing this as a second phase of implementation?

That could make sense. For issues like #454, I might need to kill govim, but I suppose that needs to be handled differently.

Or perhaps they should be handled together: if gopls dies, then govim can exit as well, and rely on the vim code to restart govim.

**Comment by myitcv** [comment]
> One trigger was switching git branches, but I don't have a repeatable repro case.

`govim` supports some level of watching changes in files not loaded in the editor. It does so by effectively pretending that the file _is_ open in the editor. But this mechanism does not support files being removed. That's a known but undocumented problem:

* largely because I couldn't see how to implement delete (shame on me for neither creating an issue, not documenting the short-coming or following up on the LSP protocol point)
* the real solution is https://github.com/myitcv/govim/issues/187 which _can_ support this

I've raised https://github.com/myitcv/govim/issues/465 to capture confirming proper delete support whilst we await #187 and the associated `gopls` support.

**Comment by myitcv** [comment]
> For issues like #454, I might need to kill govim, but I suppose that needs to be handled differently.

Killing `govim`/`govim` dying is a different kettle of fish entirely because the restart needs to happen from within the Vim [shim](https://github.com/myitcv/govim/blob/master/plugin/govim.vim).  Not given it much thought: it might be easy, might well be tricky because of the various autocommands, listeners that will be active.

**Comment by cespare** [comment]
I'm running into some gopls crash again and it's quite unfortunate that this hard-freezes vim in such a way that it has to be force-killed leaving a bunch of swap files around and blowing a large crater in my flow.

**Comment by myitcv** [comment]
Thanks for raising https://github.com/golang/go/issues/36637 off the back of this.

Given the underlying problem you encountered, what you're actually asking for here is https://github.com/govim/govim/issues/99, which is itself dependent on https://github.com/vim/vim/issues/4224. 

I'll comment over there now to see if there is anything we can do here.

**Comment by cespare** [comment]
@myitcv I think it's both #99 and this one, right? If vim didn't freeze, that would be easier to recover from, but I'd still have to restart vim because govim would no longer be working since gopls has died.

**Comment by cespare** [comment]
In fact, even without #99 if govim could recover from gopls exiting it seems like that would be enough. (I originally filed #453 for that but you wanted to combine with this one. The way this issue is titled makes it seem to be specifically about intentionally killing gopls, not crash recovery, though.)

**Comment by myitcv** [comment]
> @myitcv I think it's both #99 and this one, right? If vim didn't freeze, that would be easier to recover from, but I'd still have to restart vim because govim would no longer be working since gopls has died.

Ah yes, you are 100% right. 

> even without #99 if govim could recover from gopls exiting it seems like that would be enough

Again, correct!

Not disagreeing with your logic, but if `govim` did automatically recover from `gopls` crashing we would arguably have less issues like this raised - these issues are immensely useful, so thank you for raising and following up. Such a state would worry me somewhat - for a long time people who use VSCode just became used to firing the key combination for "restart language server" when things went wrong. Bugs go unnoticed that way. But I realise this has to be balanced with the practical aspect of disrupting people's workflows.

That said, I'd be happy if someone wanted to work on making `govim` more resilient here.

**Comment by mvdan** [comment]
I found another use case for this. If I'm heavily refactoring a codebase, including moving entire Go files between packages, `gopls` will spin my CPU at every turn trying to recompile everything. And I do need vim to continuously modify said files to fix build failures etc.

One answer could be "well, don't use your IDE-like editor for this, use a dumber editor for it". Which I sort of agree with. I want to use my regular vim setup, *except* govim/gopls.

So I'm still looking forward to a way to use vim without govim. Perhaps something like `export GOVIM_DISABLE=true`, so within that shell session, any following vim executions don't run govim at all. Right now, I can only get this behavior by completely uninstalling govim from `~/.vim`, which is cumbersome - and later I have to remember to re-enable it, too.

**Comment by myitcv** [comment]
Per our offline chat, I think it does make sense to support something like `GOVIM_DISABLE` for this case of refactoring. Principally because not all plugin managers have an easy way of disabling a plugin, but also because it is at least consistent across plugin managers (and saves the end user having to define such a thing for themselves).

**Comment by mvdan** [comment]
Indeed. I'll hopefully send a small PR to do just that - disable the plugin via an env var. It does not guarantee that gopls won't be running, as it might have been started earlier/elsewhere, but that's fine.

**Comment by myitcv** [comment]
> Indeed. I'll hopefully send a small PR to do just that - disable the plugin via an env var. It does not guarantee that gopls won't be running, as it might have been started earlier/elsewhere, but that's fine.

Indeed. But I think users of remote/daemon mode will probably be conscious of that in any case.
